### PR TITLE
Reconcile Ray Workers when VolumeMounts change

### PR DESCRIPTION
**Pull request by peterghaddad**
## Why are these changes needed?

Ensures Ray Pods are up-to-date when volumeMounts change in the cluster manifest. This needs to be utilized with the flag: ```-forced-cluster-upgrade```

## Related issue number

#944 

## Checks

When using the `-forced-cluster-upgrade` argument with the Ray Operator. The following manual steps have taken place ensuring new functionality works as intended:

1. Create small Ray Cluster utilizing example manifest.
2. When the KubeRay operator creates the Ray Cluster, the K8s API server attaches volumeMounts to the head and worker pods. Ensure this does not trigger reconciliation.
```
        - name: kube-api-access-sgtdfd
          readOnly: true
```
3. Update the `live` Ray Cluster manifest and attach a new VolumeMount to the head pod. Ensure this causes reconciliation of the head node.
4. Update the `live` Ray Cluster manifest and attach a new VolumeMount to the worker pod. Ensure this causes reconciliation of the worker node.

- [x] I've made sure the tests are passing. 
- Testing Strategy
   - [x] Unit tests
   - [x] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by peterghaddad** [comment]
Similar to https://github.com/ray-project/kuberay/issues/527

**Comment by peterghaddad** [comment]
@DmitriGekhtman This isn't 100% ready for review just yet, but think this would be good for you to be one of the reviewers if that's okay.

**Comment by DmitriGekhtman** [comment]
Sure, keep @kevin85421 in the loop as well!

File: ray-operator/Dockerfile [hunk 1]
```diff
@@ -1,6 +1,24 @@
 # Build the manager binary
 FROM golang:1.17 as builder
 
+# DELETE
```
Comment (kevin85421) [inline ray-operator/Dockerfile:4]: What's this update for?

File: ray-operator/apis/ray/v1alpha1/zz_generated.deepcopy.go [hunk 2]
```diff
@@ -6,7 +6,7 @@
 package v1alpha1
 
 import (
-	v1 "k8s.io/api/core/v1"
+	"k8s.io/api/core/v1"
```
Comment (kevin85421) [inline ray-operator/apis/ray/v1alpha1/zz_generated.deepcopy.go:9]: The CI linter will report error for this change.

File: ray-operator/controllers/ray/raycluster_controller.go [hunk 3]
```diff
@@ -92,6 +92,8 @@ type RayClusterReconciler struct {
 func (r *RayClusterReconciler) Reconcile(ctx context.Context, request ctrl.Request) (ctrl.Result, error) {
 	var err error
 
+	r.Log.Info("Hello world")
```
Comment (kevin85421) [inline ray-operator/controllers/ray/raycluster_controller.go:95]: Remove this.

File: ray-operator/controllers/ray/utils/util.go [hunk 4]
```diff
@@ -346,6 +346,9 @@ func PodNotMatchingTemplate(pod corev1.Pod, template corev1.PodTemplateSpec) boo
 					// resource entries do not match
 					return true
 				}
+				if compareVolumeMounts(container1.VolumeMounts, container2.VolumeMounts) {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/utils/util.go:349]: I think we should `return true` only when the volume mounts are different.
```suggestion
				if compareVolumeMounts(container1.VolumeMounts, container2.VolumeMounts) {
```

File: ray-operator/controllers/ray/utils/util.go [hunk 5]
```diff
@@ -429,3 +432,17 @@ func GenerateJsonHash(obj interface{}) (string, error) {
 
 	return hashStr, nil
 }
+
+func compareVolumeMounts(m1 []corev1.VolumeMount, m2 []corev1.VolumeMount) bool {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/utils/util.go:436]: The logic is weird for me. I guess this function returns `true` when volume mounts are all the same, but the logic looks like the function returns `true` when one of the volume mount is the same.

**Review by kevin85421 — COMMENTED** [review]
Thank you for the contribution! Would you mind adding some unit tests for the function? In addition, a detailed script for manually testing is good. I will also test this PR by myself.

**Comment by peterghaddad** [comment]
@kevin85421 @DmitriGekhtman This is ready for review.

File: ray-operator/controllers/ray/utils/util.go (hunk 4 above)
Comment (peterghaddad) [reply ray-operator/controllers/ray/utils/util.go]: Function no longer exists.

File: ray-operator/apis/ray/v1alpha1/zz_generated.deepcopy.go (hunk 2 above)
Comment (peterghaddad) [reply ray-operator/apis/ray/v1alpha1/zz_generated.deepcopy.go]: Thanks for this catch! Reverted.

File: ray-operator/controllers/ray/utils/util_test.go [hunk 6]
```diff
@@ -176,6 +176,16 @@ func TestReconcile_CheckNeedRemoveOldPod(t *testing.T) {
 					Image:   "rayproject/autoscaler",
 					Command: []string{"python"},
 					Args:    []string{"/opt/code.py"},
+					Env: []corev1.EnvVar{
```
Comment (kevin85421) [inline ray-operator/controllers/ray/utils/util_test.go:179]: This seems to be unrelated to this PR?

File: ray-operator/controllers/ray/utils/util_test.go [hunk 7]
```diff
@@ -238,6 +248,16 @@ func TestReconcile_CheckNeedRemoveOldPod(t *testing.T) {
 					Image:   "rayproject/autoscaler",
 					Command: []string{"echo"},
 					Args:    []string{"Hello Ray"},
+					Env: []corev1.EnvVar{
```
Comment (kevin85421) [inline ray-operator/controllers/ray/utils/util_test.go:251]: This seems to be unrelated to this PR?

File: ray-operator/controllers/ray/utils/util.go [hunk 8]
```diff
@@ -346,6 +347,10 @@ func PodNotMatchingTemplate(pod corev1.Pod, template corev1.PodTemplateSpec) boo
 					// resource entries do not match
 					return true
 				}
+				if !equality.Semantic.DeepDerivative(container1.VolumeMounts, container2.VolumeMounts) {
```
Comment (kevin85421) [inline ray-operator/controllers/ray/utils/util.go:350]: Why do we use `DeepDerivative` rather than `DeepEqual`?

**Review by kevin85421 — COMMENTED** [review]
Thank you for the update! Would you mind adding some tests in [util_test.go](https://github.com/ray-project/kuberay/blob/master/ray-operator/controllers/ray/utils/util_test.go)?

File: ray-operator/controllers/ray/utils/util.go (hunk 8 above)
Comment (peterghaddad) [reply ray-operator/controllers/ray/utils/util.go:350]: DeepEqual causes issues when default volumeMounts are mounted via the Kubernetes cluster. When using DeepEqual, I noticed pod would continuously reconcile.

**Comment by peterghaddad** [comment]
@kevin85421 Thanks for taking a look! Added in a test specifically aimed at testing the volumeMounts and changed the envVar to volumeMounts.

File: ray-operator/controllers/ray/utils/util.go (hunk 8 above)
Comment (peterghaddad) [reply ray-operator/controllers/ray/utils/util.go:350]: "I also have the problem using reflect.DeepEqual or equality.Semantic.DeepEqual because some fields are set with default non-zero values by some controller (like ImagePullPolicy, RestartPolicy, and so on) if these fields are not set by the operator" - https://github.com/kubernetes-sigs/kubebuilder/issues/592

Comment (kevin85421) [reply ray-operator/controllers/ray/utils/util.go:350]: https://github.com/kubernetes-sigs/kubebuilder/issues/592#issuecomment-768113991
https://github.com/kubernetes-sigs/kubebuilder/issues/592#issuecomment-779365927

`DeepDerivative` seems to be risky. Would you mind explaining the behavior of this function and why does it work in our case?

Comment (peterghaddad) [reply ray-operator/controllers/ray/utils/util.go:350]: Hey @kevin85421, I think those comments are actually what we want. Kubernetes generates empty or nil values, and using DeepEqual will cause problems since it will detect a drift when it shouldn't..

```
DeepDerivative is similar to DeepEqual except that unset fields in a1 are ignored (not compared). This allows us to focus on the fields that matter to the semantic comparison.
```

Comment (kevin85421) [reply ray-operator/controllers/ray/utils/util.go:350]: > DeepDerivative is similar to DeepEqual except that unset fields in a1 are ignored (not compared). This allows us to focus on the fields that matter to the semantic comparison.

I knew this. My point is: are there any situations where the worker update could be triggered unintentionally? I have this concern because:

(1) `DeepDeriative` is much more complex than what we want to compare. ([code](https://github.com/kubernetes/apimachinery/blob/master/third_party/forked/golang/reflect/deep_equal.go#L292))
(2) Which fields will be updated by Kubernetes APIServer? Will the changes trigger the worker update unintentionally?

File: ray-operator/controllers/ray/utils/util_test.go [hunk 9]
```diff
@@ -176,6 +182,16 @@ func TestReconcile_CheckNeedRemoveOldPod(t *testing.T) {
 					Image:   "rayproject/autoscaler",
 					Command: []string{"python"},
 					Args:    []string{"/opt/code.py"},
+					Env: []corev1.EnvVar{
```
Comment (kevin85421) [inline ray-operator/controllers/ray/utils/util_test.go:185]: Why do we need to add `Env` here? `Env` seems to be unused in the function `PodNotMatchingTemplate`.

File: ray-operator/controllers/ray/utils/util_test.go [hunk 10]
```diff
@@ -159,6 +159,12 @@ func TestReconcile_CheckNeedRemoveOldPod(t *testing.T) {
 							},
 						},
 					},
+					VolumeMounts: []corev1.VolumeMount{
```
Comment (kevin85421) [inline ray-operator/controllers/ray/utils/util_test.go:162]: Remove this. This will make L203 fail.

**Review by kevin85421 — COMMENTED** [review]
Have you ever ran the unit tests on your local environment? You can check this [doc](https://github.com/ray-project/kuberay/blob/master/ray-operator/DEVELOPMENT.md) for more details.

File: ray-operator/controllers/ray/utils/util.go (hunk 8 above)
Comment (peterghaddad) [reply ray-operator/controllers/ray/utils/util.go:350]: Hi Kevin. I believe we have the same concern. Using DeepDeriative for comparing volumeMounts does not trigger unintentionally. 

1. Is is more complex, but it seems that the complexity comes for ensuring fields are properly compared based off data type.
2. The common example is empty volumeMounts and volumes are added via the APIServer causing workers to update unintentionally.

File: ray-operator/controllers/ray/utils/util_test.go (hunk 10 above)
Comment (peterghaddad) [reply ray-operator/controllers/ray/utils/util_test.go]: Thanks for pointing it out. I have been running and building the tests locally each time. Think a few of my changes got reverted, but we are good now. Thanks!

**Comment by peterghaddad** [comment]
@kevin85421 Thanks for all the review! The tests have been passing locally, and thought of a couple other cases to add. Let me know if you have anything else.

**Comment by peterghaddad** [comment]
@kevin85421 wanted to politely follow up.

**Review by kevin85421 — COMMENTED** [review]
Hi, I am busy with the release v0.5.0 these weeks. Will get back to you when the release process finishes. I do not review the PR again, but something can be improved:

(1) Please provide a detailed test script in the PR description outlining how you tested this PR. Currently, KubeRay CI does not have any integration tests for `-forced-cluster-upgrade`, so I will review the related PRs carefully and test them myself. Please ensure that you thoroughly understand the behavior of both the Kubernetes API server and the `DeepDerivative` function. Both of these components make me feel risky to merge this PR. (https://github.com/ray-project/kuberay/pull/945/files#r1139782405)

(2) Can we make the test in `util_test.go` become a separate unit test?

Thanks!

**Comment by peterghaddad** [comment]
@kevin85421 I addressed the above. Thank you for taking another look!

**Comment by peterghaddad** [comment]
Hey @kevin85421 wanted to follow up on this. I appreciate the time!
